sqlglot/diff.py

```diff
@@ -0,0 +1,162 @@
+    Returns:
+        the list of Insert, Remove and Keep objects for each node in the source and the target expression trees.
+        This list represents a sequence of steps needed to transform the source expression tree into the target one.
+    """
+    expr_list_src = _expr_to_list(expr_src)
+    expr_list_tgt = _expr_to_list(expr_tgt)
+    return _myers_diff(expr_list_src, expr_list_tgt)
+
+
+def _myers_diff(expr_list_src, expr_list_tgt):
+    front = {1: (0, [])}
+
+    for d in range(0, len(expr_list_src) + len(expr_list_tgt) + 1):
+        for k in range(-d, d + 1, 2):
+            go_down = k == -d or (k != d and front[k - 1][0] < front[k + 1][0])
+
+            if go_down:
+                old_x, history = front[k + 1]
+                x = old_x
+            else:
+                old_x, history = front[k - 1]
+                x = old_x + 1
+            y = x - k
+
+            history = history[:]
+
+            if 1 <= y <= len(expr_list_tgt) and go_down:
+                history.append(Insert(expr_list_tgt[y - 1]))
+            elif 1 <= x <= len(expr_list_src):
+                history.append(Remove(expr_list_src[x - 1]))
+
+            while (
+                x < len(expr_list_src)
+                and y < len(expr_list_tgt)
+                and _is_expr_unchanged(expr_list_src[x], expr_list_tgt[y])
+            ):
+                history.append(Keep(expr_list_src[x], expr_list_tgt[y]))
+                x += 1
+                y += 1
+
+            if x >= len(expr_list_src) and y >= len(expr_list_tgt):
+                return history
+
+            front[k] = x, history
+
+    raise DiffError("Unexpected state")
+
+
+def _is_expr_unchanged(expr_src, expr_tgt):
+    equals_exprs = [Literal, Column, Table, TableAlias]
+    this_equals_exprs = [DataType, Anonymous]
+
+    if any(_all_same_type(t, expr_src, expr_tgt) for t in equals_exprs):
+        return expr_src == expr_tgt
+
+    if any(_all_same_type(t, expr_src, expr_tgt) for t in this_equals_exprs):
+        return expr_src.this == expr_tgt.this
+
+    return type(expr_src) is type(expr_tgt)
+
+
+EXCLUDED_EXPR_TYPES = (Identifier,)
+
+
+def _expr_to_list(expr):
+    return [
+        n[0]
+        for n in _normalize_args_order(expr.copy()).walk()
+        if not isinstance(n[0], EXCLUDED_EXPR_TYPES)
+    ]
+
+
+def _normalize_args_order(expr):
+    for arg in _expression_only_args(expr):
+        _normalize_args_order(arg)
+
+    if isinstance(expr, (Group, In, Select, With)):
+        for k, a in expr.args.items():
+            if isinstance(a, list) and all(
+                isinstance(a_item, Expression) for a_item in a
+            ):
+                expr.args[k] = sorted(a, key=Expression.sql)
+    elif isinstance(expr, Connector):
+        args = [expr.this, expr.args.get("expression")]
+        args = sorted(args, key=Expression.sql)
+        expr.args["this"], expr.args["expression"] = args
+
+    return expr
+
+
+def _expression_only_args(expr):
```

Comment:
i wonder if unnest_operands will help you, it gives you the added benefit of getting rid of parenthesis

Reply:
https://github.com/tobymao/sqlglot/blob/main/sqlglot/expressions.py#L243

As far as I can tell `unnest_operands` doesn't handle arguments that are themselves lists (eg. `expressions`, etc.). Or am I missing something?

Reply:
I actually expect it to fail if the list argument is encountered.

Reply:
yep, and it does!
```
>>> parse_one("a IN (1, 2)").unnest_operands()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/izeigerman/workspace/sqlglot/sqlglot/expressions.py", line 243, in unnest_operands
    return tuple(arg.unnest() for arg in self.args.values() if arg)
  File "/Users/izeigerman/workspace/sqlglot/sqlglot/expressions.py", line 243, in <genexpr>
    return tuple(arg.unnest() for arg in self.args.values() if arg)
AttributeError: 'list' object has no attribute 'unnest'
```

Reply:
gotcha